packages/zoe/src/zoe.js

```diff
@@ -285,34 +288,58 @@ const makeZoe = (additionalEndowments = {}) => {
       const instanceHandle = harden({});
       const contractFacet = makeContractFacet(instanceHandle);
 
-      const makeContractInstance = issuerRecords => {
-        const terms = {
-          ...userProvidedTerms,
-          issuers: issuerRecords.map(record => record.issuer),
-        };
+      const firstCapASCII = /^[A-Z][a-zA-Z0-9_$]*$/;
+      const { roles } = userProvidedTerms;
 
-        const instanceRecord = harden({
-          installationHandle,
-          publicAPI: undefined,
-          issuers: terms.issuers,
-          terms,
-        });
+      // `getOwnPropertyNames` returns all the non-symbol properties
+      // (both enumerable and non-enumerable).
+      const roleNames = Object.getOwnPropertyNames(roles);
+      // We sort to get a deterministic order that is not based on key
+      // insertion order or object creation order.
+      roleNames.sort();
 
-        instanceTable.create(instanceRecord, instanceHandle);
-        return Promise.resolve(
-          installation.makeContract(contractFacet, terms),
-        ).then(value => {
-          const { invite, publicAPI } = value;
-          instanceTable.update(instanceHandle, { publicAPI });
-          return invite;
-        });
-      };
+      // Insist that there are no symbol properties.
+      assert(
+        Object.getOwnPropertySymbols(roles).length === 0,
+        details`no symbol properties allowed`,
+      );
+
+      // Insist all key characters are ascii and keys start with a
+      // capital letter.
+      assert(
+        roleNames.every(name => firstCapASCII.test(name)),
+        details`roleNames must be ascii and must start with a capital letter.`,
+      );
+
+      const issuers = roleNames.map(name => roles[name]);
 
       // The issuers may not have been seen before, so we must wait for
       // the issuer records to be available synchronously
       return issuerTable
-        .getPromiseForIssuerRecords(userProvidedTerms.issuers)
-        .then(makeContractInstance);
+        .getPromiseForIssuerRecords(issuers)
+        .then(issuerRecords => {
```

Comment:
I found makeContractInstance() more readable pulled out as a named function. Is there a reason to move it inline?

Reply:
Good point. Before it had to be in a separate function or as an anonymous callback because it was a step in a change of promise.thens. Since I didn't need that anymore I moved it inline, but yeah, that separation might be more readable.